src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -805,11 +808,16 @@ private void DoBind(EndPoint endPointSnapshot, Internals.SocketAddress socketAdd
                 UpdateStatusAfterSocketErrorAndThrowException(errorCode);
             }
 
-            if (_rightEndPoint == null)
+            if (!IsWildcardEndPoint(_rightEndPoint))
             {
-                // Save a copy of the EndPoint so we can use it for Create().
-                _rightEndPoint = endPointSnapshot;
+                _localEndPoint = _rightEndPoint;
```

Comment:
This is an optimization unrelated to the PR. We can assign `_localEndPoint` with the endpoint provided by the user.

Reply:
Do we know why is existing code so aggressive with cloning/recreating EndPoints instead of working with the user-provided values?

Isn't it needed for correctness or consistency? I just realized that after fixing #1482 in #39313, now a user can alter `socket.LocalEndPoint` (for example by saying `((IPEndPoint)socket.LocalEndPoint).Port = someUnrelatedPort`), which is meaningless, and won't have any effect, but probably not something we actually wanted to enable on our API.

/cc @geoffkizer

(I wish these types were immutable ...)

Reply:
It's because the types are mutable. So in theory a user could modify them in weird ways if we don't make a copy.

> (I wish these types were immutable ...)

Me too.

One option might be to make them "freezable"; once frozen, we can trust it won't change and we don't need to make a copy then.

But I'm not sure that really adds much value; it would depend on the user calling Freeze or whatever themselves, and of course no one is doing that today.

Reply:
So in .NET Core we decided to stop caring about these weird side effects for perf?

Reply:
No, not as far as I'm aware.

Reply:
> So in .NET Core we decided to stop caring about these side effects for perf?

You're talking about the caching added in https://github.com/dotnet/runtime/issues/1482, right?

Reply:
Yes I mean that. I think we forgot to consider that `EndPoint`s are mutable types. I'm afraid if you cache a user-provided `IPEndPoint` here, that may lead to further inconsistencies.

> So in .NET Core we decided to stop caring about these side effects for perf?

> No, not as far as I'm aware.

@geoffkizer does this mean that we should consider the behavior introduced in #39313 as a bug? (Easy to fix by returning a copy of `_localEndPoint` from the property.)